Add support for Android Auto *(season 2)*

Stypox opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
Exposes media controls and playback status on Android Auto, as well as a basic browsing interface, showing playlists, history, and supporting basic searches.

This PR supersedes #9592, and the important differences lie in these commits: https://github.com/TeamNewPipe/NewPipe/pull/12044/commits/b764ad33c44e80371e79f3b64edcb3f819e0adfa, https://github.com/TeamNewPipe/NewPipe/pull/12044/commits/6eb2920f489656722790355821a764274712f9e4, https://github.com/TeamNewPipe/NewPipe/pull/12044/commits/dd29b7393ade2ba9fe72db56c85005ff23333db1. Moreover, the `MediaBrowserConnector` was split into two mostly independent classes (`MediaBrowserImpl` and `MediaBrowserPlaybackPreparer`) which also have less confusing names. You can tell my new commits apart from what was already there in #9592 by looking at the commit Co-Authors. 

I tested with `desktop-head-unit` (also used for the screenshots below) and there do not seem to be player issues anymore (in particular, the player closes correctly in all ways it can be closed and never leaved behind an empty notification). I also tested a few other player use-cases and the behavior seems to be the same as current NewPipe.

#### Issues, bugs, and missing features

* [x] ~~When closing the player from its mini view, playback continues. It should be stopped instead.~~ Fixed with the player changes
* [x] ~~When starting playback on Android Auto then opening the app, the mini player doesn't appear and isn't synchronized with the current state when "opened" (opening the details page of a content with autoplay disabled, pressing the play button plays the content on the details page)~~ Fixed in https://github.com/TeamNewPipe/NewPipe/commit/6558794d265ba8f20958c1e3e1644bcfce22773f
* [x] ~~Remote playlists are limited to their first page~~ - This is unsolvable since [Android Auto does not support pagination](https://stackoverflow.com/questions/74057148/how-does-android-auto-pagination-work-with-media3), but if you start playing a playlist (e.g. by clicking on a specific item) then the next pages will load in the queue.
* [x] ~~when starting the app then connecting to an Desktop Head Unit. The DHU was blocked on the loading step happening before showing playback controls normally, the app was playing the audio in the meantime then the app crashed.~~ Fixed with the player changes
* [x] ~~Integration with voice search~~ - voice search from within the app works, but search from the google assistant (e.g., "play <song> on NewPipe")  only supports applications from the Google Play Store.

#### Future enhancements

* Show progress bars for played items and for the one currently playing, if applicable (see https://developer.android.com/training/cars/media?hl=en#browse-progress-bar for the latter).
* Add more content to Android Auto.
* Items (playlists and streams) are not updated in Android Auto when they are in the app.
* Non square thumbnails seems to forced to the square format in the miniplayer commands if you enable the setting of the Desktop Head Unit.

#### Before/After Screenshots/Screen Record
<table>
  <tr>
    <th>Player</th>
    <th>Queue</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/b028d04e-7baf-4ff5-b66b-6ba2d0e7ed9f"/></td>
    <td><img src="https://github.com/user-attachments/assets/e12b1434-a58e-4654-b2f5-3445c791b475"/></td>
  </tr>
  <tr>
    <th>Home (bookmarked playlists)</th>
    <th>Home (history)</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/a6e5b349-ae4c-4f88-b2bc-9daccb5d9276"/></td>
    <td><img src="https://github.com/user-attachments/assets/2fed1086-2230-4f4d-bc2f-e304aaab10da"/></td>
  </tr>
  <tr>
    <th>A local playlist</th>
    <th>Search screen</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/f9e2a79a-4403-4e57-8a87-4883c923cff5"/></td>
    <td><img src="https://github.com/user-attachments/assets/cf10ba84-3c11-49ac-8c66-d67c9c4141e6"/></td>
  </tr>
</table>

#### Fixes the following issue(s)
<!-- Prefix issues with "Fixes" so that GitHub closes them when the PR is merged (note that each "Fixes #" should be in its own item). Also add any other relevant links. -->
- Fixes #1758

#### APK testing
<!-- Use a new, meaningfully named branch. The name is used as a suffix for the app ID to allow installing and testing multiple versions of NewPipe, e.g. "commentfix", if your PR implements a bugfix for comments. (No names like "patch-0" and "feature-1".)  -->
<!-- Remove the following line if you directly link the APK created by the CI pipeline. Directly linking is preferred if you need to let users test.-->
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

haggaie commented:
Thanks @Stypox for this rewrite. I had a look at the code, and it looks good to me. I also tried testing it (so far only with the desktop head unit) and it seems to work well. I saw that it fixed the issues you originally raised e.g., with closing the player on the phone after starting it on android auto. At one attempt restarting playback didn't work, but in all other attempts it worked fine.

Another thing I noticed (though I'm not sure if this wasn't happening before) is that if the player is paused, and I select a different item in Android Auto, in some cases it plays a couple of seconds from the old song before switching to the new one. Could this be related to the fix of onPrepare?

Stypox commented:
> Another thing I noticed (though I'm not sure if this wasn't happening before) is that if the player is paused, and I select a different item in Android Auto, in some cases it plays a couple of seconds from the old song before switching to the new one. Could this be related to the fix of onPrepare?

I can't reproduce, but I have noticed this too in the current version of NewPipe (before this PR). I doubt it is caused by onPrepare, since I just rewired `onPrepare()` to call `exoPlayer.prepare()`, as would have happened if no `PlaybackPreparer` was used.

haggaie commented:
Hi, I noticed a crash this morning that might be related to the changes in the service lifecycle. If I understand correctly, the video fragment's onDestroy triggers player service shutdown, which then attempts accessing the same fragment:

## Exception
* __User Action:__ ui error
* __Request:__ ACRA report
* __Content Country:__ IL
* __Content Language:__ en-IL
* __App Language:__ en_IL_#u-fw-sun-mu-celsius
* __Service:__ none
* __Timestamp:__ 2025-02-18T06:55:03.431+02:00
* __Package:__ org.schabi.newpipe.debug.androidauto
* __Service:__ none
* __Version:__ 0.27.6
* __OS:__ Linux Android 15 - 35
<details><summary><b>Crash log </b></summary><p>

```
java.lang.RuntimeException: Unable to destroy activity {org.schabi.newpipe.debug.androidauto/org.schabi.newpipe.MainActivity}: java.lang.NullPointerException: Attempt to read from field 'android.widget.ImageButton org.schabi.newpipe.databinding.FragmentVideoDetailBinding.overlayPlayPauseButton' on a null object reference in method 'void org.schabi.newpipe.fragments.detail.VideoDetailFragment.setOverlayPlayPauseImage(boolean)'
	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:6044)
	at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:6076)
	at android.app.servertransaction.DestroyActivityItem.execute(DestroyActivityItem.java:52)
	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:63)
	at android.app.servertransaction.TransactionExecutor.executeLifecycleItem(TransactionExecutor.java:169)
	at android.app.servertransaction.TransactionExecutor.executeTransactionItems(TransactionExecutor.java:101)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:80)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2719)
	at android.os.Handler.dispatchMessage(Handler.java:109)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8787)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:871)
Caused by: java.lang.NullPointerException: Attempt to read from field 'android.widget.ImageButton org.schabi.newpipe.databinding.FragmentVideoDetailBinding.overlayPlayPauseButton' on a null object reference in method 'void org.schabi.newpipe.fragments.detail.VideoDetailFragment.setOverlayPlayPauseImage(boolean)'
	at org.schabi.newpipe.fragments.detail.VideoDetailFragment.setOverlayPlayPauseImage(VideoDetailFragment.java:2414)
	at org.schabi.newpipe.fragments.detail.VideoDetailFragment.onServiceStopped(VideoDetailFragment.java:1851)
	at org.schabi.newpipe.player.helper.PlayerHolder$1.onServiceStopped(PlayerHolder.java:312)
	at org.schabi.newpipe.player.Player.stopActivityBinding(Player.java:2038)
	at org.schabi.newpipe.player.Player.destroy(Player.java:608)
	at org.schabi.newpipe.player.PlayerService.cleanup(PlayerService.java:207)
	at org.schabi.newpipe.player.PlayerService.destroyPlayerAndStopService(PlayerService.java:231)
	at org.schabi.newpipe.player.helper.PlayerHolder.stopService(PlayerHolder.java:145)
	at org.schabi.newpipe.fragments.detail.VideoDetailFragment.onDestroy(VideoDetailFragment.java:396)
	at androidx.fragment.app.Fragment.performDestroy(Fragment.java:3376)
	at androidx.fragment.app.FragmentStateManager.destroy(FragmentStateManager.java:812)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:346)
	at androidx.fragment.app.SpecialEffectsController$FragmentStateManagerOperation.complete(SpecialEffectsController.kt:664)
	at androidx.fragment.app.SpecialEffectsController$Operation.cancel(SpecialEffectsController.kt:507)
	at androidx.fragment.app.SpecialEffectsController.forceCompleteAllOperations(SpecialEffectsController.kt:293)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3037)
	at androidx.fragment.app.FragmentManager.dispatchDestroy(FragmentManager.java:2988)
	at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:346)
	at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:258)
	at androidx.appcompat.app.AppCompatActivity.onDestroy(AppCompatActivity.java:283)
	at org.schabi.newpipe.MainActivity.onDestroy(MainActivity.java:460)
	at android.app.Activity.performDestroy(Activity.java:9310)
	at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1563)
	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:6031)
	... 14 more

```
</details>
<hr>

Note that this is from the first version of this PR, so maybe it was already fixed?

notbasa commented:
Got this while testing new android auto implementation, but I have no idea if this is in any way related to the AA code.
User perceived effect is just that the playing seems to stop, and user needs to repress the play button for it to continue.

## Exception* __User Action:__ play stream* __Request:__ Player error[type=ERROR_CODE_IO_BAD_HTTP_STATUS] occurred while playing https://www.youtube.com/watch?v=13m5-0EOwIM* __Content Country:__ US* __Content Language:__ en-US* __App Language:__ en_US* __Service:__ YouTube* __Timestamp:__ 2025-02-17T20:36:13.019+02:00* __Package:__ org.schabi.newpipe.debug.androidauto* __Service:__ YouTube* __Version:__ 0.27.6* __OS:__ Linux Nokia/TheThing_00EEA/TTG_sprout:14/UKQ1.231003.002/00WW_4_170:user/release-keys 14 - 34<details>

<summary><b>Crash log </b></summary><p>
```com.google.android.exoplayer2.ExoPlaybackException: Source error	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:644)	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:616)	at android.os.Handler.dispatchMessage(Handler.java:102)	at android.os.Looper.loopOnce(Looper.java:205)	at android.os.Looper.loop(Looper.java:294)	at android.os.HandlerThread.run(HandlerThread.java:67)Caused by: com.google.android.exoplayer2.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 403	at org.schabi.newpipe.player.datasource.YoutubeHttpDataSource.open(YoutubeHttpDataSource.java:425)	at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:263)	at com.google.android.exoplayer2.upstream.TeeDataSource.open(TeeDataSource.java:52)	at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource(CacheDataSource.java:796)	at com.google.android.exoplayer2.upstream.cache.CacheDataSource.open(CacheDataSource.java:609)	at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:84)	at com.google.android.exoplayer2.source.chunk.ContainerMediaChunk.load(ContainerMediaChunk.java:124)	at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:412)	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)	at java.lang.Thread.run(Thread.java:1012)
```</details><hr>

Stypox commented:
> the player displays the Prev and next arrows on the UI but they don't cause any action when pressed. Ideally they should change the song to next/prev on the history list or they should be grayed out if this is not feasible for some reason.

This is not under our control, as starting from Android 13 (i.e. Tiramisu) the previous, next and play/pause actions are no longer customizable and are set by the system. Do other apps behave differently when there is only one song in queue?

> Hi, I noticed a crash this morning that might be related to the changes in the service lifecycle

Thanks, fixed in https://github.com/TeamNewPipe/NewPipe/pull/12044/commits/126f4b0e30e321646537951b21f67a562ec9d6db

> Got this while testing new android auto implementation, but I have no idea if this is in any way related to the AA code.

This is unrelated, see https://github.com/TeamNewPipe/NewPipe/issues/11803

---

I also pushed commits https://github.com/TeamNewPipe/NewPipe/pull/12044/commits/a7a7dc53631579ea71baed0f097c0ef8638ae07b and https://github.com/TeamNewPipe/NewPipe/pull/12044/commits/6558794d265ba8f20958c1e3e1644bcfce22773f which update the code after the assumptions about how the player service is stopped and reused have changed (i.e. the service can now outlive the player). 

> - When starting playback on Android Auto then opening the app, the mini player doesn't appear and isn't synchronized with the current state when "opened" (opening the details page of a content with autoplay disabled, pressing the play button plays the content on the details page)

Fixed in https://github.com/TeamNewPipe/NewPipe/pull/12044/commits/6558794d265ba8f20958c1e3e1644bcfce22773f. Note that this behavior was buggy before too, e.g. when the player was started from RouterActivity (i.e. the share menu) while MainActivity was not open.

haggaie commented:
Hi,

I tested the previous version (73f2cfa I guess) on my car as well, and it seemed to work well. I've also tested the new version (a7a7dc5) on the desktop-head-unit, and it also works great.

I finally understood the comment about the square desktop-head-unit aspect ratio, and I was able to reproduce it, but I noticed that other apps also exhibit the same behavior. Could it be a problem in how Android Auto supports this aspect ratio?

harirah1 commented:
![IMG_20250220_165745](https://github.com/user-attachments/assets/7b530872-4140-4eda-b376-b286bc2ce6b2)
![IMG_20250220_165700](https://github.com/user-attachments/assets/c25d626e-88f0-4b5c-884a-132369347eb0)
Hello, this is the first time I'm posting a comment! I tested the version I downloaded on 19.02.25. It works perfectly. No errors to report. The only thing I could mention is that I can't select random mode directly on the car screen. So I put my playlist and random mode via the phone and then I play my music via the steering wheel interface without any problems. This is the biggest thing I wanted: to avoid playing music on the phone. Successful thing. Thank you!

Stypox commented:
Thank you all for testing! @haggaie could you explicitly approve the code, if you have reviewed it all? @Isira-Seneviratne and @litetex could you also give a review, especially on the player lifecycle, if you have some time? Thanks :-)

> Could it be a problem in how Android Auto supports this aspect ratio?

Yeah it might be. I wouldn't consider it a blocker anyway.

AndresPuglia98 commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.java:
```diff
@@ -177,9 +299,55 @@ public static class LocalBinder extends Binder {
+     */
+    public void setPlayerListener(@Nullable final Consumer<Player> listener) {
+        this.onPlayerStartedOrStopped = listener;
+        if (listener != null) {
+            if (player == null) {
+                listener.accept(null);
+            } else {
+                listener.accept(player);
+            }
         }
```
Hi, I'm leaving a small suggestion here:
```suggestion
    if (listener == null) {
        return;
    }
    listener.accept(player);
```

Stypox replied:
I think the reason it's written like this is that I wanted to make it explicit that the listener might be called with a null value, but probably a short comment would explain it better. Thanks for the catch ;-)

Stypox replied:
```suggestion
        if (listener != null) {
            // if there is no player, then `null` will be sent here, to make sure the state is synced
            listener.accept(player);
        }
```

haggaie reviewed (approved):
Thanks @Stypox. I reviewed the code, and it seemed fine to me. I also managed to test it with my car after my previous comment.

Stypox commented:
> but in my car, android auto doesnt display newpipe app on android auto display 1 time of 2
> i have to unplug and replug my phone to android auto

That's strange, it's up to the Android system to decide when to query apps so I don't know if we can do anything about it.

harirah1 commented:
Hello,

Yesterday, I had a Google update. After that, NewPipe Auto was no longer recognized in Android Auto. I tried reinstalling it, but I got the message "App not installed." This is strange because external sources are allowed for downloads.

Today, I updated a few apps, including Android Auto, and NewPipe Auto became accessible again in the settings. However, when I tried to reinstall the app via Check, CL, etc., I once again got the "App not installed" message. I don’t know why. In any case, the version I installed on 19.02 is still there. I just hope the car recognizes it.

If you need more information, feel free to ask—I’m not sure how else I can help!

Have a great day!

![Screenshot_20250301_095829_com google android packageinstaller](https://github.com/user-attachments/assets/d613c038-fd6a-4465-b871-b0bd64b22ddb)
![Screenshot_20250301_093052_com android vending](https://github.com/user-attachments/assets/7c563f6b-13c5-4728-89f6-5838c35962a0)

Occhioverde commented:
> I tried reinstalling it, but I got the message "App not installed." This is strange because external sources are allowed for downloads.

@harirah1 This is caused by the fact that debug APKs generated by the CI are (apparently, someone who knows better can correct me if I'm wrong) signed with a different key each time. In order to update/reinstall this build of NewPipe you first need to uninstall it. This obviously doesn't apply to the release version, so if you have it installed alongside the `android-auto` build you can keep it.

andry23jsv commented:
I noticed that on latest version, after using newpipe on android auto, the app is remaining active in background. 
On previous version this not happens.

Stypox commented:
> I noticed that on latest version, after using newpipe on android auto, the app is remaining active in background.

Yes, that's normal. The system is keeping the service active in the background, but the service is not actually doing anything and should get killed by the system after some time. It will only happen after using NewPipe on Android Auto anyway (or other media browser queries). NewPipe's code already tries to shutdown the background service when the player is closed, but the system refuses to shut it down.

oleg-d commented:
I installed the apk from the CI but my android auto is not showing New Pipe in the launcher. Did you guys do anything apart from install the apk to get it running? This is on a Pixel 9.

ready2code31 commented:
Hello, the app not working anymore after removing it and reinstalling the app

ready2code31 commented:
> > but in my car, android auto doesnt display newpipe app on android auto display 1 time of 2
> > i have to unplug and replug my phone to android auto
> 
> That's strange, it's up to the Android system to decide when to query apps so I don't know if we can do anything about it.

I think i found a solution to this
you have to install the app-debug.apk using adb to set app provider to com.android.vending (play store)
https://medium.com/@pixplicity/setting-the-install-vendor-of-an-app-7d7deacb01ee
and then each time, you plug your phone to the car
the app display on android auto well
thx devs for this pull, it works great

Stypox commented:
I'm merging this, so it ends up in nightly APKs and gets tested by more people. Thank you again `@haggaie`!

Stypox commented:
https://github.com/TeamNewPipe/NewPipe-nightly/releases

abesqaured commented:
Hello, I am unable to start NewPipe on Android auto.

Stypox commented:
@abesqaured which version are you on? This PR is still not in any official version but only in nightlies

sanchiro82 commented:
Thank you so much for your contribution, it's incredible. It would be great to include your subscription timeline.

Stypox commented:
Could someone test #12412 on a real car, just to make sure the changes did not break anything?
